File: coding/src/com/algos/li15_tree/Bt_L100_SameTree.java
package  com.algos.li15_tree;

import  com.algos.li30_model.TreeNode;

public class Bt_L100_SameTree {
    public static void main(String[] args) {
        Bt_L100_SameTree obj = new Bt_L100_SameTree();

        TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);
        
        System.out.println(obj.isSameTree(root, root));
    }

    // Problem: https://leetcode.com/problems/same-tree/ 
    // Idea: Root values should be same and left and right sub trees should be same trees.
    // time: O(n)
    // space: O(1)
    public boolean isSameTree(TreeNode root1, TreeNode root2) {
		if(root1 == null || root2 == null)
			return root1 == root2;
		
        return (root1.val == root2.val) && isSameTree(root1.left, root2.left) && isSameTree(root1.right, root2.right);
	}
}

File: dsa/topics/10_trees/bt/python/bt_l100__same_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Problem: https://leetcode.com/problems/same-tree/
# Idea: Root values should be same and left and right subtrees should also be same.
# Time: O(n)
# Space: O(1) ignoring recursion stack

def is_same_tree(root1, root2):
    if not root1 or not root2:
        return root1 is root2

    return (root1.val == root2.val and 
            is_same_tree(root1.left, root2.left) and 
            is_same_tree(root1.right, root2.right))


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

print(is_same_tree(root, root))  # Expected: True
